pinot-common/src/main/java/org/apache/pinot/common/cursors/AbstractResponseStore.java

```diff
@@ -0,0 +1,186 @@
+   */
+  protected abstract ResultTable readResultTable(String requestId)
+      throws Exception;
+
+  protected abstract boolean deleteResponseImpl(String requestId)
+      throws Exception;
+
+  /**
+   * Stores the response in the store. @link{CursorResponse} and @link{ResultTable} are stored separately.
+   * @param response Response to be stored
+   * @throws Exception Thrown if there is any error while storing the response.
+   */
+  public void storeResponse(CursorResponse response)
+      throws Exception {
+    String requestId = response.getRequestId();
+
+    try {
+      long bytesWritten = writeResultTable(requestId, response.getResultTable());
+
+      // Remove the resultTable from the response as it is serialized in a data file.
+      response.setResultTable(null);
+      response.setBytesWritten(bytesWritten);
+      writeResponse(requestId, response);
+      _brokerMetrics.addMeteredGlobalValue(BrokerMeter.CURSOR_RESULT_STORE_SIZE, bytesWritten);
+    } catch (Exception e) {
+      _brokerMetrics.addMeteredGlobalValue(BrokerMeter.CURSOR_WRITE_EXCEPTION, 1);
+      deleteResponse(requestId);
+      throw e;
+    }
+  }
+
+  /**
+   * Reads the response from the store and populates it with a slice of the @link{ResultTable}
+   * @param requestId Request ID of the query
+   * @param offset Offset of the result slice
+   * @param numRows Number of rows required in the slice
+   * @return A CursorResponse with a slice of the @link{ResultTable}
+   * @throws Exception Thrown if there is any error during the operation.
+   */
+  public CursorResponse handleCursorRequest(String requestId, int offset, int numRows)
+      throws Exception {
+
+    CursorResponse response;
+    ResultTable resultTable;
+
+    try {
+      response = readResponse(requestId);
+    } catch (Exception e) {
+      _brokerMetrics.addMeteredGlobalValue(BrokerMeter.CURSOR_READ_EXCEPTION, 1);
+      throw e;
+    }
+
+    int totalTableRows = response.getNumRowsResultSet();
+
+    if (totalTableRows == 0 && offset == 0) {
+      // If sum records is 0, then result set is empty.
+      response.setResultTable(null);
+      response.setOffset(0);
+      response.setNumRows(0);
+      return response;
+    } else if (offset >= totalTableRows) {
+      throw new RuntimeException("Offset " + offset + " is greater than totalRecords " + totalTableRows);
+    }
+
+    long fetchStartTime = System.currentTimeMillis();
+    try {
+      resultTable = readResultTable(requestId);
+    } catch (Exception e) {
+      _brokerMetrics.addMeteredGlobalValue(BrokerMeter.CURSOR_READ_EXCEPTION, 1);
+      throw e;
+    }
+
+    int sliceEnd = offset + numRows;
+    if (sliceEnd > totalTableRows) {
+      sliceEnd = totalTableRows;
+      numRows = sliceEnd - offset;
+    }
+
+    response.setResultTable(
+        new ResultTable(resultTable.getDataSchema(), resultTable.getRows().subList(offset, sliceEnd)));
```

Comment:
Does that mean every time a read request will fetch the whole result table into memory first and then pick the page from the memory?

Reply:
This is dependent on the implementation of the plugin. The default FsResponseStore that uses the file system does read the whole result set into memory.

Reply:
This implementation does read the whole result table into memory. Does LinkedIn implementation support seeking within the result table ? Do you have suggestions on how can the interface be changed to also support seek ?